pkg/events/raw/raw.go

```diff
@@ -0,0 +1,207 @@
+}
+
+type Stream struct {
+	js jetstream.Stream
+}
+
+func FromConfig(ctx context.Context, name string, cfg Config) (*Stream, error) {
+	var s *Stream
+	b := backoff.NewExponentialBackOff()
+
+	connect := func() error {
+		var tlsConf *tls.Config
+		if cfg.EnableTLS {
+			var rootCAPool *x509.CertPool
+			if cfg.TLSRootCACertificate != "" {
+				rootCrtFile, err := os.Open(cfg.TLSRootCACertificate)
+				if err != nil {
+					return err
+				}
+
+				rootCAPool, err = newCertPoolFromPEM(rootCrtFile)
+				if err != nil {
+					return err
+				}
+				cfg.TLSInsecure = false
+			}
+
+			tlsConf = &tls.Config{
+				MinVersion:         tls.VersionTLS12,
+				InsecureSkipVerify: cfg.TLSInsecure,
+				RootCAs:            rootCAPool,
+			}
+		}
+
+		nopts := nats.GetDefaultOptions()
+		nopts.Name = name
+		if tlsConf != nil {
+			nopts.Secure = true
+			nopts.TLSConfig = tlsConf
+		}
+
+		if len(cfg.Endpoint) > 0 {
+			nopts.Servers = []string{cfg.Endpoint}
+		}
+
+		if cfg.AuthUsername != "" && cfg.AuthPassword != "" {
+			nopts.User = cfg.AuthUsername
+			nopts.Password = cfg.AuthPassword
+		}
+
+		conn, err := nopts.Connect()
+		if err != nil {
+			return err
+		}
+
+		jsConn, err := jetstream.New(conn)
+		if err != nil {
+			return err
+		}
+
+		js, err := jsConn.Stream(ctx, events.MainQueueName)
+		if err != nil {
+			return err
+		}
+
+		s = &Stream{
+			js: js,
+		}
+		return nil
+	}
+	err := backoff.Retry(connect, b)
+	if err != nil {
+		return s, errors.Wrap(err, "could not connect to nats jetstream")
+	}
+	return s, nil
+}
+
+func (s *Stream) Consume(group string, evs ...events.Unmarshaller) (<-chan Event, error) {
+	c, err := s.consumeRaw(group)
+	if err != nil {
+		return nil, err
+	}
+
+	registeredEvents := map[string]events.Unmarshaller{}
+	for _, e := range evs {
+		typ := reflect.TypeOf(e)
+		registeredEvents[typ.String()] = e
+	}
+
+	outchan := make(chan Event)
+	go func() {
+		for {
+			e := <-c
+			eventType := e.Metadata[events.MetadatakeyEventType]
+			ev, ok := registeredEvents[eventType]
+			if !ok {
+				_ = e.msg.Ack() // Discard. We are not interested in this event type
+				continue
+			}
+
+			event, err := ev.Unmarshal(e.Payload)
+			if err != nil {
+				continue
+			}
+
+			outchan <- Event{
+				Event: events.Event{
+					Type:        eventType,
+					ID:          e.Metadata[events.MetadatakeyEventID],
+					TraceParent: e.Metadata[events.MetadatakeyTraceParent],
+					InitiatorID: e.Metadata[events.MetadatakeyInitiatorID],
+					Event:       event,
+				},
+				msg: e.msg,
+			}
+		}
+	}()
+	return outchan, nil
+}
+
+func (s *Stream) consumeRaw(group string) (<-chan RawEvent, error) {
+	consumer, err := s.js.CreateOrUpdateConsumer(context.Background(), jetstream.ConsumerConfig{
+		Durable:       group,
+		DeliverPolicy: jetstream.DeliverNewPolicy,
+		AckPolicy:     jetstream.AckExplicitPolicy, // Require manual acknowledgment
+		MaxAckPending: 10000,                       // Maximum number of unacknowledged messages
```

Comment:
Isn't that somewhat excessive for a default? Especially since you change the "search" implementation it immediately ack additional events for spaces that are sitting already in the debouncer?

In the end I fear we need to make this configurable per consumer same for the `AckWait` settinging.

Reply:
I agree that it's very unlikely to ever reach that limit in case of the search service, but it's a different story for postprocessing, where events are only acknowledged after they have been persisted in the storage. I easily got over 1000 pending acks during my tests and I wouldn't want the nats queue to slow down the postprocessing.